code/datums/components/usb_port.dm

```diff
@@ -0,0 +1,163 @@
+	usb_cable_ref = null
+
+	return ..()
+
+/datum/component/usb_port/proc/unregister_circuit_signals()
+	if (isnull(attached_circuit))
+		return
+
+	UnregisterSignal(attached_circuit, list(
+		COMSIG_CIRCUIT_SHELL_REMOVED,
+		COMSIG_PARENT_QDELETING,
+	))
+
+	var/shell = attached_circuit.shell
+	if (!isnull(shell))
+		UnregisterSignal(shell, COMSIG_MOVABLE_MOVED)
+		UnregisterSignal(shell, COMSIG_PARENT_EXAMINE)
+
+/datum/component/usb_port/proc/create_circuit_components(obj/item/integrated_circuit/circuitboard)
+	var/created_circuit_components = list()
+
+	for(var/circuit_component_type in circuit_component_types)
+		var/obj/item/circuit_component/circuit_component = new circuit_component_type(parent)
+		circuit_component.insertable_through_shells = TRUE
+		circuitboard.add_component(circuit_component)
+		created_circuit_components += circuit_component
+		RegisterSignal(circuit_component, COMSIG_CIRCUIT_COMPONENT_REMOVED, .proc/on_circuit_component_removed)
+
+	return created_circuit_components
+
+/datum/component/usb_port/proc/on_examine(datum/source, mob/user, list/examine_text)
+	SIGNAL_HANDLER
+
+	if (isnull(attached_circuit))
+		examine_text += "<span class='notice'>There is a USB port on the front.</span>"
+	else
+		examine_text += "<span class='notice'>[attached_circuit.shell || attached_circuit] is connected to [parent.p_them()] by a USB port.</span>"
+
+/datum/component/usb_port/proc/on_examine_shell(datum/source, mob/user, list/examine_text)
+	SIGNAL_HANDLER
+
+	examine_text += "<span class='notice'>[source.p_they(TRUE)] [source.p_are()] attached to [parent] with a USB cable.</span>"
+
+/datum/component/usb_port/proc/on_atom_usb_cable_try_attach(datum/source, obj/item/usb_cable/connecting_cable, mob/user)
+	SIGNAL_HANDLER
+
+	var/atom/atom_parent = parent
+
+	if (!isnull(attached_circuit))
+		atom_parent.balloon_alert(user, "usb already connected")
+		return COMSIG_CANCEL_USB_CABLE_ATTACK
+
+	if (isnull(connecting_cable.attached_circuit))
+		connecting_cable.balloon_alert(user, "connect to a shell first")
+		return COMSIG_CANCEL_USB_CABLE_ATTACK
+
+	if (!IN_GIVEN_RANGE(connecting_cable.attached_circuit, parent, USB_CABLE_MAX_RANGE))
+		connecting_cable.balloon_alert(user, "too far away")
+		return COMSIG_CANCEL_USB_CABLE_ATTACK
+
+	usb_cable_ref = WEAKREF(connecting_cable)
+	attached_circuit = connecting_cable.attached_circuit
+
+	connecting_cable.forceMove(attached_circuit)
+	circuit_components = create_circuit_components(attached_circuit)
+	attached_circuit.interact(user)
+
+	usb_cable_beam = atom_parent.Beam(attached_circuit.shell, "usb_cable_beam", 'icons/obj/wiremod.dmi')
+
+	RegisterSignal(attached_circuit, COMSIG_CIRCUIT_SHELL_REMOVED, .proc/on_circuit_shell_removed)
+	RegisterSignal(attached_circuit, COMSIG_PARENT_QDELETING, .proc/on_circuit_deleting)
+	RegisterSignal(attached_circuit.shell, COMSIG_MOVABLE_MOVED, .proc/on_moved)
+	RegisterSignal(attached_circuit.shell, COMSIG_PARENT_EXAMINE, .proc/on_examine_shell)
```

Comment:
shell can be null if the attached circuit doesn't have a shell initially.

Reply:
The USB cable itself requires that it be in a shell before you connect.